File: erlang/collatz-conjecture/src/collatz_conjecture.erl
-module(collatz_conjecture).

-export([steps/1]).


steps(N) when N =< 0 ->
  error(badarg);
steps(N) ->
  steps(N, 0).

steps(1, S) ->
   S;
steps(N, S) when N rem 2 == 0 ->
  steps(N div 2, S + 1);
steps(N, S) ->
  steps(N * 3 + 1, S + 1).

File: elixir/collatz-conjecture/lib/collatz_conjecture.ex
defmodule CollatzConjecture do
  @doc """
  calc/1 takes an integer and returns the number of steps required to get the
  number to 1 when following the rules:
    - if number is odd, multiply with 3 and add 1
    - if number is even, divide by 2
  """
  def calc(input) when input <= 0, do: raise FunctionClauseError
  def calc(input), do: calc(input, 0)

  def calc(1, steps), do: steps

  def calc(input, steps) when rem(input, 2) == 0 do
    calc(div(input, 2), steps + 1)
  end

  def calc(input, steps) when rem(input, 2) == 1 do
    calc(input * 3 + 1, steps + 1)
  end
end
